xprof/xprof.rb.in

```diff
@@ -563,34 +602,34 @@ end
 # Start, Stop lttng, amd do the on-node analsysis
 def trace_and_on_node_processing(usr_argv)
   # All masters set a future global barrier
-  syncd = Sync_daemon.new
-
-  # Load Tracers and APILoaders Lib
-  backends, h = env_tracers
-
-  # All ranks need to set the LLTTNG_HOME env
-  # so they can have access to the daemon
-  ENV['LTTNG_HOME'] = lttng_home_dir
-  # Only local master spawn LTTNG daemon and start session
-  setup_lttng(backends) if mpi_local_master?
-  syncd.local_barrier('waiting_for_lttng_setup')
-  # Launch User Command
-  launch_usr_bin(h, usr_argv)
-
-  # We need to be sure that all the local ranks are finished
-  # before the local master stops the lttng session
-  syncd.local_barrier('waiting_for_application_ending')
-  return unless mpi_local_master?
-
-  teardown_lttng
-  # Preprocess trace
-  on_node_processing(backends, syncd)
+  Sync_daemon.open do |syncd|
+    # Load Tracers and APILoaders Lib
+    backends, h = env_tracers
+
+    # All ranks need to set the LLTTNG_HOME env
+    # so they can have access to the daemon
+    ENV['LTTNG_HOME'] = lttng_home_dir
+    # Only local master spawn LTTNG daemon and start session
+    setup_lttng(backends) if mpi_local_master?
+    syncd.local_barrier('waiting_for_lttng_setup')
+    # Launch User Command
+    launch_usr_bin(h, usr_argv)
+
+    # We need to be sure that all the local ranks are finished
+    # before the local master stops the lttng session
+    syncd.local_barrier('waiting_for_application_ending')
+    return unless mpi_local_master?
+
+    teardown_lttng
+    # Preprocess trace
+    on_node_processing(backends, syncd)
+  end
```

Comment:
I think I get the idea, some ruby magic to make sure finalize is called no matter what, but not clear to me why it wouldn't be called?

Reply:
Because before, due to all the early exit (`return unless master_local?`) some specific path where not called. 
Like in Open, context manager help solve this issue. (a little bit like RAII in C++)